### ISSUE TITLE
Inconsistent handling of to_string() removal suggestion between field and method access

### ISSUE THREAD

**Issue by ecpost** [rust-lang/rust-clippy#13726]
### Summary

I have 2 values to pass to a function.  The first is borrowed (indirectly) from a field in the second, and the function needs a mutable reference of the second.  So I am making a copy (`to_string()`) of the first, to avoid the conflicting borrows.  But clippy suggests removing `to_string()`, which would result in a compiler error.

However, this suggestion only happens if the field is borrowed via a method.  If done directly with the field instead of an accessor method, clippy does not make the (invalid) suggestion.

### Lint Name

unnecessary_to_owned

### Reproducer

Here is a complete example:

```rust
struct Something {
    value: String,
    // other fields
}

impl Something {
    fn value(&self) -> &str {
        &self.value
    }
}

fn use_it(_value: &str, _thing: &mut Something) {
    // do stuff
}

fn main() {
    let mut thing = Something { value: "abc".to_string() };

    // This (using field `value` directly) is fine, does not generate a warning:
    use_it(&thing.value.to_string(), &mut thing);

    // But this (using method `value()` to read field `value`) generates a warning:
    use_it(&thing.value().to_string(), &mut thing);

    // Using the suggestion results in a compiler error because of the immutable and mutable borrows together:
    // use_it(thing.value(), &mut thing);
}
```

The first call to `use_it` (using the field directly) does NOT result in a warning.  But the second call results in:

```
warning: unnecessary use of `to_string`
  --> src/main.rs:23:12
   |
23 |     use_it(&thing.value().to_string(), &mut thing);
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `thing.value()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
   = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
```

And the third call to `use_it` with the suggested change (commented out) would result in a compiler error because of the conflicting borrows.

### Version

```text
rustc 1.81.0 (eeb90cda1 2024-09-04)
binary: rustc
commit-hash: eeb90cda1969383f56a2637cbd3037bdf598841c
commit-date: 2024-09-04
host: x86_64-apple-darwin
release: 1.81.0
LLVM version: 18.1.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by qsantos** [comment]
I have encountered this issue as well, and reduced it to:
```rust
struct S(String);
impl S {
    fn g(&mut self, _s: &str) { unimplemented!() }
    fn f(&mut self) {
        let a = &self.0;
        let b = &a[0..];
        self.g(&b.to_string()); // Clippy suggest using the form below
        //self.g(b); // does not compile
    }
}
```
With
```
rustc 1.83.0 (90b35a623 2024-11-26)
binary: rustc
commit-hash: 90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf
commit-date: 2024-11-26
host: x86_64-unknown-linux-gnu
release: 1.83.0
LLVM version: 19.1.1
```
For now, I juse use `#[allow(clippy::unnecessary_to_owned)]`.

**Comment by blyxyas** [comment]
Same issue as #14918
